grafana/opensearch-datasource
# PPL: Execute Explore PPL logs format queries through the backend

**What this PR does / why we need it**:

This executes PPL queries with the `format` set to `logs` from Explore to be executed on the backend.

To test:
- Go to Explore
- Set "Query" to `PPL`
- Set "Format" to `Logs`
- Use the "Logs" panel
- Enter your PPL query: you can find some example queries [here](https://clouddatasources.grafana.net/goto/vMOq6OmIg?orgId=1) under the "PPL Logs" row

**PPL logs queries will only run in the backend in the Explore view** in order to limit this migration's impact to dashboards. Please [let us know](https://github.com/grafana/opensearch-datasource/issues/new?assignees=&labels=datasource%2FOpenSearch%2Ctype%2Fbug&projects=&template=bug_report.md) if you observe some unwanted behaviour while executing PPL Log queries in Explore.

**Which issue(s) this PR fixes**:

Fixes #200 

**Special notes for your reviewer**:

**There are some differences between the frontend execution and backend execution**:
- Fields are sorted alphabetically. The only exceptions are the time field and the log message field. The time field is the first field, and the log message field is the second field if it is configured in the data source settings.
- Fields are no longer treated as strings, instead they are typed properly. E.g. the frontend would treat `false` as a string, but the backend will now treat is as a bool.

**Testing ad-hoc filters**:
The logs panel doesn't provide a button to apply ad-hoc filters. Instead you'll need to test this from a dashboard with the following procedure:
- Enable PPL log queries to be sent to the backend by removing `request.app === CoreApp.Explore` from line 546 in `src/datasource.ts`
- Create a new visualization
- Set query to "PPL" and format to "Logs"
- Use the "Table" panel
- Run a query (you can just leave the query field blank if you have an index configured)
- Wait until some data is returned and then hover over one of the cells in the table panel and click the "+" icon.
- You can change to the "Logs" panel
- Alternatively, you can just add an ad-hoc filter template variable.
